feat: add NetworkPolicy support

matskiv opened the pull request:
This should enable NetworPolicy synchronization with all necessary rewrites for the pod label selectors and namespace label selectors.

Fixes #155

matskiv commented:
I properly updated the code based on the recent changes in main + made this feature opt-in to allow easy installation even if the user doesn't have the permissions for network policies.
The new test will still fail, as there are at least two missing pieces here:
a) KinD cluster used for e2e tests probably needs to be configured with a different CNI plugin, one that supports network policies
b) New controller needs to be enabled in the deployments used for e2e tests. This might just be another devspace profile :thinking:

olljanat commented:
> a) KinD cluster used for e2e tests probably needs to be configured with a different CNI plugin, one that supports network policies

I see that you ended up disabling Calico now so recommend that you considering using K3s on network policy related e2e tests instead of KinD (and you can spin those with K3d) as it does support network policies without installing separate CNI.

FabianKramm commented on charts/k0s/values.yaml:
```diff
@@ -85,6 +85,9 @@ rbac:
   role:
     # This is required for basic functionality of vcluster
     create: true
+  extendedRolePermissions:
```
Please rename this to `rbac.role.extended`

FabianKramm commented on charts/k3s/values.yaml:
```diff
@@ -93,6 +93,9 @@ rbac:
   role:
     # This is required for basic functionality of vcluster
     create: true
+  extendedRolePermissions:
```
See above

FabianKramm commented on charts/k8s/values.yaml:
```diff
@@ -132,6 +132,9 @@ rbac:
   role:
     # This is required for basic functionality of vcluster
     create: true
+  extendedRolePermissions:
```
See above

FabianKramm commented on devspace.yaml:
```diff
@@ -40,6 +40,8 @@ deployments:
         rbac:
           clusterRole:
             create: true
+          extendedRolePermissions:
```
see above

FabianKramm commented on pkg/controllers/resources/networkpolicies/translate.go:
```diff
@@ -0,0 +1,128 @@
+	return newNetworkPolicy, nil
+}
+
+func (s *syncer) translateUpdate(pObj, vObj *networkingv1.NetworkPolicy) (*networkingv1.NetworkPolicy, error) {
```
no error is ever returned here, so please omit it

FabianKramm reviewed (changes requested):
@matskiv thanks for the PR! Please take a look at my comments, besides that it looks good!

matskiv commented:
@olljanat Thank you for the suggestion, I'll give it a shot :)

matskiv commented:
I addressed review comments.

As for the failing tests, after some investigation, it doesn't seem to be related to network policy changes. The failures happen due to a number of pod/exec calls which sometimes trigger yet unknown issue on the server that causes it to be temporarily unresponsive to new exec calls. I was not able to tell which component is causing this because multiple pods are marked as not ready when this happens, so it could be a problem on a lower network level... Anyway, since this is not caused by the network policy controller I will create a follow-up issue, and the tests will be skipped in CI for now.

matskiv commented:
@olljanat yes, that's the plan. Once this is merged a new beta release will be done.

FabianKramm reviewed (approved):
@matskiv LGTM! Thanks!
